### ISSUE TITLE
Chat - If url ends with slash, then ending double quote is shown as link

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#24367]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Go to https://staging.new.expensify.com/
2. Tap on a report
3. In compose, enter "om [https://www.she.com/"](https://www.she.com/%22)
"om [https://www.she.com"](https://www.she.com%22/)
4. Tap send

## Expected Result:
Irrespective of url ends with/without slash, url alone must be displayed as link and not ending double quote

## Actual Result:
If url ends with slash, then ending double quote is shown as link. If url ends without slash, then correctly url alone displayed as link

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.52.3

**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/78819774/3ea950da-b0ae-4025-84a2-90b956b72657

**Expensify/Expensify Issue URL:**

**Issue reported by:** Applause - Internal Team

**Slack conversation:**

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by greg-schroeder** [comment]
@lanitochka17 can you clarify which platforms this occurs on?

**Comment by gzqyl** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
When users input URLs in the chat application that end with a slash and then immediately close with a double quote, the application incorrectly includes the double quote as part of the URL link.
### What is the root cause of that problem?
the common lib parse links is not perfect, while it is very good enough. It does parse correct many url links with its regex rule, even already considerate of the "(" and ")", it does not include '"'.
### What changes do you think we should make in order to solve the problem?
I do not think revise the common lib is a good idea, as it is already good enough.
Indeed, app part has its own utilities function which does not depend on common lib very much, just add some regex rule, it will solve this issue, as follow, I successfully solved this issue, with or without slash, the url link is ok.
My solution does not need common lib changes, just use the common lib function, and add some extra regex rule is ok.

![Screenshot_1691805395](https://github.com/Expensify/App/assets/39096450/1b2fe83e-4ced-475b-8333-a436f68f2a61)


### What alternative solutions did you explore? (Optional)
Change the common lib directly, but I really do not suggest to solve the issue like so.

**Comment by gzqyl** [comment]
Contributor details
Expensify account email: [allen.gzqyl@gmail.com](mailto:allen.gzqyl@gmail.com)
Upwork Profile Link: https://www.upwork.com/freelancers/~01e99782a01937e92c

**Comment by ShogunFire** [comment]
I think the problem is to know what we want, do we accept quotes like this "" after the .com or not ? And do we accept escaped quotes like this: &quot; ? Because right now we are escaping the comment before testing the regexp... So I don't think we can have a different behaviour for " and for &quot. So what should be this behavior ?

**Comment by ShogunFire** [comment]
And in slack if you don't have a / &quot is not recognized either

**Comment by greg-schroeder** [comment]
What do you think about the above, @parasharrajat?

**Comment by greg-schroeder** [comment]
Sure - assigned you @situchan

**Comment by situchan** [comment]
Thanks for proposals. I don't think anyone proposed right solution to make url ending with slash work always.
This is basic testing:
<img width="173" alt="Screenshot 2023-08-21 at 4 23 35 PM" src="https://github.com/Expensify/App/assets/108292595/b5f532a9-37b8-4eb0-8b35-7a8b403aa094">

Please provide exact root cause why ending with `/` not work while without `/` works

**Comment by ShogunFire** [comment]
@situchan Because the " is transformed into &quot and with slash it is accepted as a path or argument. This is normal behaviour. I think for &quot but not for ". I explaoned the problem here

https://github.com/Expensify/App/issues/24367#issuecomment-1676467982

**Comment by Antasel** [comment]
# Proposal

### Please re-state the problem that we are trying to solve in this issue.
When users input URLs in the chat application that end with a slash and then immediately close with a double quote, the application incorrectly includes the double quote as part of the URL link.

### What is the root cause of that problem?
The root cause is that double quote is being allowed in Url.
but according to [RFC2396](https://datatracker.ietf.org/doc/html/rfc2396), double quote is not allowed. instead we should use escaped form %22 for double quote.
Backend is not allowing it as well. e.g: seding `http://test.com/test="test"` returns `http://test.com/test=`

https://github.com/Expensify/App/assets/139204153/ea19c56f-8fac-4dd3-8265-44bc37c42d54



### What changes do you think we should make in order to solve the problem?
in line https://github.com/Expensify/expensify-common/blob/main/lib/Url.js#L6, we are expected to remove 'quot'

also need some tweaks on auto test case
in line https://github.com/Expensify/expensify-common/blob/main/__tests__/ExpensiMark-HTML-test.js#L403, replace double quote with %22.
in line https://github.com/Expensify/expensify-common/blob/main/__tests__/ExpensiMark-HTML-test.js#L444, replace `&quot;` with %22.

and we can add following test case for this issue.
```
    testString = '"om https://www.she.com/"';
    resultString = '&quot;om <a href=\"https://www.she.com/\" target=\"_blank\" rel=\"noreferrer noopener\">https://www.she.com/</a>&quot;';
    expect(parser.replace(testString)).toBe(resultString);
```

**Comment by ShogunFire** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

When there is a " at the end of the url, it is recognized as part of the url

### What is the root cause of that problem?

When we transform the links in html `<a>` we use this regexp for the path of the url: https://github.com/Expensify/expensify-common/blob/7735de14112a968fd6a4f5af710d2fbaefc8809d/lib/Url.js#L7 This regexp accept quotes as a valid path, the path is only after a / that's why we don't accept quote directly after  .com for example 

### What changes do you think we should make in order to solve the problem?

We had the same exact problem with parenthesis so we can use the same way to remove the end quote if it's not preceded by another quote.

This is how we were removing the end parenthesis:
https://github.com/Expensify/expensify-common/blob/7735de14112a968fd6a4f5af710d2fbaefc8809d/lib/ExpensiMark.js#L450-L472

### What alternative solutions did you explore? (Optional)

**Comment by greg-schroeder** [comment]
Any thoughts on these updated proposals @situchan?

**Comment by lanitochka17** [comment]
@greg-schroeder Device on which the issue was found: Redmi note 10 s / Chrome

**Comment by greg-schroeder** [comment]
bump @situchan

**Comment by greg-schroeder** [comment]
bumped in c+ room

**Comment by situchan** [comment]
updating today

**Comment by greg-schroeder** [comment]
Bump @situchan

**Comment by greg-schroeder** [comment]
Bump @situchan

**Comment by situchan** [comment]
For contributors: can you please share screenshot with these links after applying your solution?
```
https://www.she.com/path?test="123"
https://www.she.com/path?test="123/"
"https://www.she.com/path?test="123"
"https://www.she.com/path?test="123/"
https://www.she.com/path?test=123"
https://www.she.com/path?test=123/"
https://www.she.com/path?test=/"123"
https://www.she.com/path?test=/"123/"
```

**Comment by ShogunFire** [comment]
My [proposal](https://github.com/Expensify/App/issues/24367#issuecomment-1687108530)  can adapt to what the expected behaviour will be. The truth is there is not a great way to know if the quote is part of the url or not , I was thinking to remove quotes if they are unmatched lile we will do for end parenthesis but for quotes which one to remove if they are 3. Actually I believe we should do lile slacks and remove them all as long as they are like this: `"` and not alreafy escaped: `&quot;` my solution is the only one that can do this because I am not using the regexp where the text is already escaped.

**Comment by Antasel** [comment]
@situchan Slack does not accept double quote in URL
![image](https://github.com/Expensify/App/assets/139204153/3de14b11-73d8-4124-804c-a4a7d7946c8f)

**Comment by situchan** [comment]
> @situchan Slack does not accept double quote in URL

@Antasel I know. Did you check https://github.com/Expensify/App/issues/24367#issuecomment-1714528065?

**Comment by Antasel** [comment]
@situchan yes, checked.
I hope you check video again on my [proposal.](https://github.com/Expensify/App/issues/24367#issuecomment-1686408161) Backend does not accept double quote, but just the escaped form %22.
Checking of slack and the video, I am sure we should follow like Slack for this issue

**Comment by situchan** [comment]
I like @Antasel's [proposal](https://github.com/Expensify/App/issues/24367#issuecomment-1686408161).
I agree double quote should not be included in url. If included, escape it!
Also consistent with backend logic.
🎀 👀 🎀 C+ reviewed

<img width="695" alt="Screenshot 2023-09-12 at 9 53 23 AM" src="https://github.com/Expensify/App/assets/108292595/cd5e4796-b2ed-4f46-a469-5b867325caf0">


Before fix:
<img width="695" alt="Screenshot 2023-09-12 at 9 52 20 AM" src="https://github.com/Expensify/App/assets/108292595/ab2d1abd-3163-43ff-abb4-aa84f6e78d34">

After fix:
<img width="695" alt="Screenshot 2023-09-12 at 9 48 19 AM" src="https://github.com/Expensify/App/assets/108292595/8b7d5007-be44-43f6-af1b-8aa9e3caabe8">

Slack:
<img width="695" alt="Screenshot 2023-09-12 at 9 55 27 AM" src="https://github.com/Expensify/App/assets/108292595/1bad2ca5-44ba-4e80-b98d-080e5e597a65">

**Comment by ShogunFire** [comment]
@situchan I know this was not a proposal but @Antasel proposal is basically what I was commenting here, https://github.com/Expensify/App/issues/24367#issuecomment-1676468774

I didn't go that way because I thought it could create regression by not only removing `"` but the form "&quot;"

**Comment by situchan** [comment]
@ShogunFire I saw that comment but @Antasel provided explanation enough why that solution is ideal.
The main reason why I accepted that solution is it's consistent with backend logic.

**Comment by gzqyl** [comment]
@ShogunFire indeed states some root cause, just remove "quote" in addEscapedChar is not enough to solve this issue perfectly.
test the string 
![截屏2023-09-12 20 41 25](https://github.com/Expensify/App/assets/39096450/f888b21b-01c7-45ca-9a8c-cced34f06dd5)
you could find that it stops at "=", it is not same as slack.
I know this url is a little stranger, but for the user, he may write it as the image shows, so the url maybe not perfect as we expected with a "?".
And just remove the "quote" still can not solve the url like this 
![截屏2023-09-12 20 46 47](https://github.com/Expensify/App/assets/39096450/6f10c4d7-805d-4ca3-b839-dae39783c8f1)
the result is not ideal, and it is a little out of our expectation.
![截屏2023-09-12 20 47 45](https://github.com/Expensify/App/assets/39096450/d698f8bb-9a93-4c4a-ae40-0bf84cd03e5c)
We have no good way to deal with "&quot", if the user will never input this, then the issue will be much easier to solve.

**Comment by puneetlath** [comment]
I agree with bringing this in line with how the back-end does it. Let's make sure we add tests for the various scenarios.

**Comment by Antasel** [comment]
@puneetlath  @situchan 
The PR https://github.com/Expensify/expensify-common/pull/574 for expensify-common repo is ready for review

**Comment by gzqyl** [comment]
@ShogunFire how do you think of this solution which is not perfect? we know it will be still wrong when the string contains "&quot;", and it does not like slack, or even GitHub, if we just need to fix "/", there will be many methods to fix it, but fix this issue perfect like slack or GitHub, it will be a little hard, just remove "quot" is too simple, and can only fix the issue at some extent.

**Comment by gzqyl** [comment]
check the images, the bug still exists
![截屏2023-09-14 10 26 28](https://github.com/Expensify/App/assets/39096450/37cca954-f2c6-4741-b7ab-6f45f69b3bcd)
![截屏2023-09-14 10 27 50](https://github.com/Expensify/App/assets/39096450/dcb9e796-c509-4801-8a54-287a1b06df05)
![截屏2023-09-14 10 28 59](https://github.com/Expensify/App/assets/39096450/a37bf939-9cae-443e-bd8a-88fd3d1a707f)

**Comment by situchan** [comment]
@gzqyl we don't need to follow slack, github

**Comment by greg-schroeder** [comment]
Work continues on linked PR

**Comment by puneetlath** [comment]
@Antasel I've merged the expensify-common PR. Can you please go ahead and open a PR in App to incorporate the changes?

**Comment by Antasel** [comment]
> @Antasel I've merged the expensify-common PR. Can you please go ahead and open a PR in App to incorporate the changes?

ok, I am in

**Comment by agilejune** [comment]
@puneetlath 
We sorry to say about @Antasel deceased by sudden stroke. we can't believe it yet.
unfortunately our team can't access via his github account, if expensify engineer team allow us to use other account, it will be benefit to us and his family..
if can't, please move forward by internal team.

**Comment by greg-schroeder** [comment]
Hey @situchan - given the situation outlined [here](https://github.com/Expensify/App/issues/24367#issuecomment-1730501614), do you think you'd be able to step in to help out with Puneet's request [here](https://github.com/Expensify/App/issues/24367#issuecomment-1724305774)?

**Comment by situchan** [comment]
Sorry to hear that.
@puneetlath should I raise app PR?
Actually I am not too worried about app PR as I think expensify-common package version will be bumped anytime from any other PRs.

**Comment by puneetlath** [comment]
@situchan yes, if you don't mind that would be great. We can add urgency bonus for the trouble.

**Comment by greg-schroeder** [comment]
Hi @agilejune - Ah, wow, that is really sad news, thanks for letting us know. This is definitely a delicate situation, but we have pretty hard and fast rules when it comes to paying out issues. As it stands, we can only pay the contributor's account directly and won't be able to redirect any payments or anything like that. Also I'm not sure if there's a ton of action we can take to help with GitHub account acesss.

Have you had a chance to contact Upwork or Github to inform them of the situation? It's likely they may have processes in place to help folks gain access to accounts that belong to deceased colleagues/family members.

**Comment by agilejune** [comment]
@greg-schroeder 
I contacted his family. and there is policy to collect funds for a deceased freelancer’s estate on upwork.
I am sure you would release milestone for @Antasel 's work.
then his family can get funds via other payment method from upwork

**Comment by greg-schroeder** [comment]
@agilejune thanks. I will just process the issue normally and will let you know as we move towards payment

**Comment by greg-schroeder** [comment]
Sorry for delay, was OOO. Processing today.

**Comment by greg-schroeder** [comment]
Payment summary:

Contributor: @Antasel $1000
C+: @situchan $1000

Paid out @agilejune fyi

**Comment by situchan** [comment]
@greg-schroeder can you please check https://github.com/Expensify/App/issues/24367#issuecomment-1732968580?

**Comment by greg-schroeder** [comment]
will fire regression test using automation steps
